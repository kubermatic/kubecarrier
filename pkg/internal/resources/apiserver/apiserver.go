/*
Copyright 2019 The KubeCarrier Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package apiserver

import (
	"fmt"
	"strings"

	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"sigs.k8s.io/kustomize/v3/pkg/image"
	"sigs.k8s.io/kustomize/v3/pkg/types"
	"sigs.k8s.io/yaml"

	operatorv1alpha1 "github.com/kubermatic/kubecarrier/pkg/apis/operator/v1alpha1"
	"github.com/kubermatic/kubecarrier/pkg/internal/kustomize"
	"github.com/kubermatic/kubecarrier/pkg/internal/resources/constants"
	"github.com/kubermatic/kubecarrier/pkg/internal/version"
)

// Config holds the config information to generate the KubeCarrier master controller manager setup.
type Config struct {
	// Namespace is the KubeCarrier master controller manager should be deployed into.
	Namespace string
	// Name of this KubeCarrier API object.
	Name string

	// Spec of the APIServer
	Spec operatorv1alpha1.APIServerSpec
}

var k = kustomize.NewDefaultKustomize()

// Manifests generate all required manifests for the API Server
func Manifests(c Config) ([]unstructured.Unstructured, error) {
	v := version.Get()
	kc := k.ForHTTP(vfs)
	if err := kc.MkLayer("man", types.Kustomization{
		Namespace: c.Namespace,
		Images: []image.Image{
			{
				Name:   "quay.io/kubecarrier/apiserver",
				NewTag: v.Version,
			},
		},
		PatchesStrategicMerge: []types.PatchStrategicMerge{
			"manager_env_patch.yaml",
		},
		Resources: []string{"../default"},
	}); err != nil {
		return nil, fmt.Errorf("cannot mkdir: %w", err)
	}

	extraArgs := make([]string, 0)
	if len(c.Spec.OIDC.RequiredClaims) > 0 {
		rclaims := make([]string, 0)
		for k, v := range c.Spec.OIDC.RequiredClaims {
			rclaims = append(rclaims, fmt.Sprintf("%s=%s", k, v))
		}
		extraArgs = append(extraArgs, "--oidc-required-claim="+strings.Join(rclaims, ","))
	}

	// Patch environment
	// Note:
	// we are not using *appsv1.Deployment here,
	// because some fields will be defaulted to empty and
	// interfere with the strategic merge patch of kustomize.
	managerEnv := map[string]interface{}{
		"apiVersion": "apps/v1",
		"kind":       "Deployment",
		"metadata": map[string]string{
			"name":      "manager",
			"namespace": "system",
		},
		"spec": map[string]interface{}{
			"template": map[string]interface{}{
				"spec": map[string]interface{}{
					"containers": []map[string]interface{}{
						{
							"name": "manager",
							"args": append([]string{
								"--address=$(API_SERVER_ADDR)",
								"--tls-cert-file=$(API_SERVER_TLS_CERT_FILE)",
								"--tls-private-key-file=$(API_SERVER_TLS_PRIVATE_KEY_FILE)",
								"--oidc-issuer-url=$(API_SERVER_OIDC_ISSUER_URL)",
								"--oidc-client-id=$(API_SERVER_OIDC_CLIENT_ID)",
								"--oidc-ca-file=$(API_SERVER_OIDC_CA_FILE)",
								"--oidc-username-claim=$(API_SERVER_OIDC_USERNAME_CLAIM)",
								"--oidc-username-prefix=$(API_SERVER_OIDC_USERNAME_PREFIX)",
								"--oidc-groups-claim=$(API_SERVER_OIDC_GROUPS_CLAIM)",
								"--oidc-groups-prefix=$(API_SERVER_OIDC_GROUPS_PREFIX)",
								"--oidc-signing-algs=$(API_SERVER_OIDC_SIGNING_ALGS)",
							}, extraArgs...),
							"env": []map[string]interface{}{
								{
									"name":  "API_SERVER_ADDR",
									"value": ":8443",
								},
								{
									"name":  "API_SERVER_TLS_CERT_FILE",
									"value": "/run/serving-certs/tls.crt",
								},
								{
									"name":  "API_SERVER_TLS_PRIVATE_KEY_FILE",
									"value": "/run/serving-certs/tls.key",
								},
								{
									"name":  "API_SERVER_OIDC_ISSUER_URL",
									"value": c.Spec.OIDC.IssuerURL,
								},
								{
									"name":  "API_SERVER_OIDC_CLIENT_ID",
									"value": c.Spec.OIDC.ClientID,
								},
								{
									"name":  "API_SERVER_OIDC_CA_FILE",
									"value": "/run/oidc-certs/ca.crt",
								},
								{
									"name":  "API_SERVER_OIDC_USERNAME_CLAIM",
									"value": c.Spec.OIDC.UsernameClaim,
								},
								{
									"name":  "API_SERVER_OIDC_USERNAME_PREFIX",
									"value": c.Spec.OIDC.UsernamePrefix,
								},
								{
									"name":  "API_SERVER_OIDC_GROUPS_CLAIM",
									"value": c.Spec.OIDC.GroupsClaim,
								},
								{
									"name":  "API_SERVER_OIDC_GROUPS_PREFIX",
									"value": c.Spec.OIDC.GroupsPrefix,
								},
								{
									"name":  "API_SERVER_OIDC_SIGNING_ALGS",
									"value": strings.Join(c.Spec.OIDC.SupportedSigningAlgs, ","),
								},
							},
							"volumeMounts": []map[string]interface{}{{
								"mountPath": "/run/serving-certs",
								"readyOnly": true,
								"name":      "serving-cert",
							}, {
								"mountPath": "/run/oidc-certs",
								"readyOnly": true,
								"name":      "oidc-cert",
							}},
						},
					},
					"volumes": []map[string]interface{}{{
						"name": "serving-cert",
						"secret": map[string]interface{}{
							"secretName": c.Spec.TLSSecretRef.Name,
						},
					}, {
						"name": "oidc-cert",
						"secret": map[string]interface{}{
							"secretName": c.Spec.OIDC.CertificateAuthority.Name,
						},
					}},
				},
			},
		},
	}
	managerEnvBytes, err := yaml.Marshal(managerEnv)
	if err != nil {
		return nil, fmt.Errorf("marshalling manager env patch: %w", err)
	}
	if err = kc.WriteFile("/man/manager_env_patch.yaml", managerEnvBytes); err != nil {
		return nil, fmt.Errorf("writing manager_env_patch.yaml: %w", err)
	}

	// execute kustomize
	objects, err := kc.Build("/man")
	if err != nil {
		return nil, fmt.Errorf("running kustomize build: %w", err)
	}

	for _, obj := range objects {
		labels := obj.GetLabels()
		if labels == nil {
			labels = map[string]string{}
		}
		labels[constants.NameLabel] = "apiserver"
		labels[constants.InstanceLabel] = c.Name
		labels[constants.ManagedbyLabel] = constants.ManagedbyKubeCarrierOperator
		labels[constants.VersionLabel] = v.Version
		obj.SetLabels(labels)
	}
	return objects, nil
}
