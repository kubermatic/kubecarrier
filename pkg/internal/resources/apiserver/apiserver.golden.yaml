- apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRole
  metadata:
    creationTimestamp: null
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
      kubecarrier.io/role: api-server
    name: kubecarrier-api-server-manager-role
  rules:
  - apiGroups:
    - ""
    resources:
    - secrets
    verbs:
    - get
    - list
    - watch
  - apiGroups:
    - authentication.k8s.io
    resources:
    - tokenreviews
    verbs:
    - create
  - apiGroups:
    - authorization.k8s.io
    resources:
    - subjectaccessreviews
    verbs:
    - create
  - apiGroups:
    - catalog.kubecarrier.io
    resources:
    - accounts
    verbs:
    - get
    - list
    - watch
  - apiGroups:
    - catalog.kubecarrier.io
    resources:
    - offerings
    verbs:
    - get
    - list
    - watch
  - apiGroups:
    - catalog.kubecarrier.io
    resources:
    - providers
    verbs:
    - get
    - list
    - watch
  - apiGroups:
    - catalog.kubecarrier.io
    resources:
    - regions
    verbs:
    - get
    - list
    - watch
- apiVersion: v1
  kind: ServiceAccount
  metadata:
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
      kubecarrier.io/role: api-server
    name: kubecarrier-api-server-sa
    namespace: kubecarrier-system-10
- apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  metadata:
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
      kubecarrier.io/role: api-server
    name: kubecarrier-api-server-manager-rolebinding
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: kubecarrier-api-server-manager-role
  subjects:
  - kind: ServiceAccount
    name: kubecarrier-api-server-sa
    namespace: kubecarrier-system-10
- aggregationRule:
    clusterRoleSelectors:
    - matchLabels:
        kubecarrier.io/apiserver: "true"
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRole
  metadata:
    creationTimestamp: null
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
      kubecarrier.io/role: api-server
    name: kubecarrier-api-server-agg-manager-role
  rules: []
- apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  metadata:
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
      kubecarrier.io/role: api-server
    name: kubecarrier-api-server-agg-manager-rolebinding
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: kubecarrier-api-server-agg-manager-role
  subjects:
  - kind: ServiceAccount
    name: kubecarrier-api-server-sa
    namespace: kubecarrier-system-10
- apiVersion: rbac.authorization.k8s.io/v1
  kind: Role
  metadata:
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
      kubecarrier.io/role: api-server
    name: kubecarrier-api-server-leader-election-role
    namespace: kubecarrier-system-10
  rules:
  - apiGroups:
    - ""
    resources:
    - configmaps
    verbs:
    - get
    - list
    - watch
    - create
    - update
    - patch
    - delete
  - apiGroups:
    - ""
    resources:
    - configmaps/status
    verbs:
    - get
    - update
    - patch
  - apiGroups:
    - ""
    resources:
    - events
    verbs:
    - create
- apiVersion: rbac.authorization.k8s.io/v1
  kind: RoleBinding
  metadata:
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
      kubecarrier.io/role: api-server
    name: kubecarrier-api-server-leader-election-rolebinding
    namespace: kubecarrier-system-10
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: Role
    name: kubecarrier-api-server-leader-election-role
  subjects:
  - kind: ServiceAccount
    name: kubecarrier-api-server-sa
    namespace: kubecarrier-system-10
- apiVersion: apps/v1
  kind: Deployment
  metadata:
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
      control-plane: api-server
      kubecarrier.io/role: api-server
    name: kubecarrier-api-server-manager
    namespace: kubecarrier-system-10
  spec:
    replicas: 1
    selector:
      matchLabels:
        control-plane: api-server
        kubecarrier.io/role: api-server
    strategy: {}
    template:
      metadata:
        labels:
          control-plane: api-server
          kubecarrier.io/role: api-server
      spec:
        containers:
        - args:
          - --address=$(API_SERVER_ADDR)
          - --tls-cert-file=$(API_SERVER_TLS_CERT_FILE)
          - --tls-private-key-file=$(API_SERVER_TLS_PRIVATE_KEY_FILE)
          - --authentication-mode=$(AUTHENTICATION_MODE)
          - -v=$(VERBOSE)
          - --htpasswd-secret-name=$(HTPASSWD_SECRET_NAME)
          - --oidc-issuer-url=$(API_SERVER_OIDC_ISSUER_URL)
          - --oidc-client-id=$(API_SERVER_OIDC_CLIENT_ID)
          - --oidc-ca-file=$(API_SERVER_OIDC_CA_FILE)
          - --oidc-username-claim=$(API_SERVER_OIDC_USERNAME_CLAIM)
          - --oidc-username-prefix=$(API_SERVER_OIDC_USERNAME_PREFIX)
          - --oidc-groups-claim=$(API_SERVER_OIDC_GROUPS_CLAIM)
          - --oidc-groups-prefix=$(API_SERVER_OIDC_GROUPS_PREFIX)
          - --oidc-signing-algs=$(API_SERVER_OIDC_SIGNING_ALGS)
          env:
          - name: API_SERVER_ADDR
            value: :8443
          - name: API_SERVER_TLS_CERT_FILE
            value: /run/serving-certs/tls.crt
          - name: API_SERVER_TLS_PRIVATE_KEY_FILE
            value: /run/serving-certs/tls.key
          - name: AUTHENTICATION_MODE
            value: Htpasswd,OIDC,Token,Anonymous
          - name: VERBOSE
            value: "0"
          - name: HTPASSWD_SECRET_NAME
            value: test-secret
          - name: API_SERVER_OIDC_ISSUER_URL
          - name: API_SERVER_OIDC_CLIENT_ID
          - name: API_SERVER_OIDC_CA_FILE
            value: /run/oidc-certs/ca.crt
          - name: API_SERVER_OIDC_USERNAME_CLAIM
          - name: API_SERVER_OIDC_USERNAME_PREFIX
          - name: API_SERVER_OIDC_GROUPS_CLAIM
          - name: API_SERVER_OIDC_GROUPS_PREFIX
          - name: API_SERVER_OIDC_SIGNING_ALGS
          image: quay.io/kubecarrier/apiserver:was not build properly
          livenessProbe:
            tcpSocket:
              port: https
          name: manager
          ports:
          - containerPort: 8443
            name: https
            protocol: TCP
          readinessProbe:
            tcpSocket:
              port: https
          resources:
            limits:
              cpu: 100m
              memory: 100Mi
            requests:
              cpu: 100m
              memory: 100Mi
          volumeMounts:
          - mountPath: /run/serving-certs
            name: serving-cert
            readOnly: true
          - mountPath: /run/oidc-certs
            name: oidc-cert
            readOnly: true
        serviceAccountName: kubecarrier-api-server-sa
        terminationGracePeriodSeconds: 10
        volumes:
        - name: serving-cert
          secret:
            secretName: apiserver-tls-cert
        - name: oidc-cert
          secret: {}
  status: {}
- apiVersion: v1
  kind: Service
  metadata:
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
      control-plane: api-server
      kubecarrier.io/role: api-server
    name: kubecarrier-api-server-manager
    namespace: kubecarrier-system-10
  spec:
    ports:
    - name: https
      port: 443
      protocol: TCP
      targetPort: https
    selector:
      control-plane: api-server
      kubecarrier.io/role: api-server
- apiVersion: cert-manager.io/v1alpha2
  kind: Issuer
  metadata:
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
    name: kubecarrier-apiserver-selfsigned-issuer
    namespace: kubecarrier-system-10
  spec:
    selfSigned: {}
- apiVersion: cert-manager.io/v1alpha2
  kind: Certificate
  metadata:
    labels:
      app.kubernetes.io/instance: foo
      app.kubernetes.io/managed-by: kubecarrier-operator
      app.kubernetes.io/name: apiserver
      app.kubernetes.io/version: was not build properly
    name: kubecarrier-apiserver-serving-cert
    namespace: kubecarrier-system-10
  spec:
    dnsNames:
    - kubecarrier-api-server-manager.kubecarrier-system-10.svc
    - kubecarrier-api-server-manager.kubecarrier-system-10.svc.cluster.local
    - localhost
    isCA: true
    issuerRef:
      kind: Issuer
      name: kubecarrier-apiserver-selfsigned-issuer
    secretName: apiserver-tls-cert
